app/javascript/hooks/use-history.ts

```diff
@@ -11,28 +11,36 @@ function removeEmpty<TParams>(obj: TParams) {
         return v !== null
       }
     })
-    .reduce((acc, [k, v]) => ({ ...acc, [k]: v }), {})
+    .reduce((acc, [k, v]) => ({ ...acc, [k]: v }), {}) as TParams
 }
 
 function toQuery<TParams>(params: TParams): string {
-  const state = removeEmpty(params)
-
-  return `?${stringify(decamelizeKeys(state), { arrayFormat: 'brackets' })}`
+  return `?${stringify(decamelizeKeys(params as Record<string, unknown>), {
+    arrayFormat: 'brackets',
+  })}`
 }
 
 function pushState<TParams>(params: TParams) {
   history.pushState(history.state, '', toQuery(params))
 }
 
 export function useHistory<TParams>({ pushOn }: { pushOn: TParams }): void {
+  const [history, setHistory] = useState(removeEmpty(pushOn))
   const isMounted = useRef(false)
 
   useEffect(() => {
+    const newHistory = removeEmpty(pushOn)
+
+    setHistory(newHistory)
+  }, [pushOn])
+
+  useEffect(() => {
+    console.log(history)
     if (!isMounted.current) {
       isMounted.current = true
       return
     }
 
-    pushState<TParams>(pushOn)
-  }, [pushOn])
+    pushState<TParams>(history)
+  }, [JSON.stringify(history)])
```

Comment:
@SleeplessByte Is there any downside to using a function as the dependency?

Reply:
You're not using a function as the dependency.

What are you trying to achieve?